torch/distributed/fsdp/_symbolic_trace.py

```diff
@@ -0,0 +1,241 @@
+import contextlib
+import functools
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, Generator, List, Optional, Set, Tuple
+
+import torch
+
+
+__all__ = ["TracingConfig"]
+
+
+@dataclass
+class TracingConfig:
+    """
+    Configurations used in ``ParamExecOrderWrapPolicy`` for symbolic tracing of a model.
+
+    Args:
+        tracer (torch.fx.Tracer): An instance of ``torch.fx.Tracer`` that will be used to perform symbolic
+            tracing. ``tracer`` is default to be ``torch.fx.Tracer()``, but can also be instance
+            of some child class of ``torch.fx.Tracer``. For example, one may want to use ``HFTracer``
+            for models in Transformers:
+            .. _Transformers: https://huggingface.co/docs/transformers/index
+        concrete_args (Optional[Dict[str, Any]]): Concrete arguments that should not be treated
+            as ``torch.fx.Proxy`` when tracing the forward function.
+            ``concrete_args`` allows one to partially specialize the forward function,
+            including removing control flow or data structures.
+            ``concrete_args`` is also the argument used in :meth:`~torch.fx.Tracer.trace`.
+    """
+
+    tracer: torch.fx.Tracer = torch.fx.Tracer()
+    concrete_args: Optional[Dict[str, Any]] = None
```

Comment:
should we enforce concrete_args to be None? when there is dynamic branches depending on input values, the torch.fx tracing approach for non-recursive wrapping will not work as expected any way?